Gems/Atom/Asset/Shader/Code/Source/Editor/ShaderVariantListBuilder.cpp

```diff
@@ -0,0 +1,586 @@
+            GetBuilderSettingsFromRegistry(enableHashCompare, suddenChangeWaitTime);
+            AZ_Warning(ShaderVariantListBuilderName, enableHashCompare, "Hash Compare has been disabled by user from registry key: %s."
+                "All variants will be considered new to build.", EnableHashCompareRegistryKey);
+
+            // As we calculate hashes for each variant, we are going to calculate the hash
+            // of only the optionValues part and this will help us find variants that share the same
+            // content but have different StableIds. If two variants with different StableId has the same content this will be an error.
+            AZStd::unordered_map<size_t, uint32_t> hashToStableIdMap;
+            HashedVariantListSourceData hashedVariantList;
+            hashedVariantList.m_shaderFilePath = shaderVariantList.m_shaderFilePath;
+            for (const auto& variantInfo : shaderVariantList.m_shaderVariants)
+            {
+                size_t optionValuesHash = HashedVariantListSourceData::HashedVariantInfo::HashCombineShaderOptionValues(0, variantInfo.m_options);
+                const auto itor = hashToStableIdMap.find(optionValuesHash);
+                if (itor != hashToStableIdMap.end())
+                {
+                    AZ_Error(ShaderVariantListBuilderName, false, "StableId [%u] has the same option values as StableId[%u].\n", variantInfo.m_stableId, itor->second);
+                    response.m_resultCode = AssetBuilderSDK::ProcessJobResult_Failed;
+                    return;
+                }
+                hashToStableIdMap.emplace(optionValuesHash, variantInfo.m_stableId);
+            
+                HashedVariantListSourceData::HashedVariantInfo hashedVariantInfo;
+                hashedVariantInfo.m_isNew = true;
+                hashedVariantInfo.m_variantInfo = variantInfo;
+                hashedVariantInfo.CalculateHash(optionValuesHash);
+                hashedVariantList.m_hashedVariants.emplace_back(AZStd::move(hashedVariantInfo));
+            }
+
+            AZStd::chrono::system_clock::time_point now = AZStd::chrono::system_clock::now();
+            hashedVariantList.m_timeStamp = now.time_since_epoch().count();
+
+            if (enableHashCompare)
+            {
+                // Time to load the previously generated *.hashedvariantlist file and compare hashes for each variantInfo.
+                // hashes that match, we'll flag as `hashedVariantInfo.m_isNew = false;` so the ShaderVariantAsset does not rebuild those.
+                AZStd::string previousHashedVariantListPath = GetAbsolutePathOfIntermediateAsset(variantListFullPath);
+                AZ_Trace(ShaderVariantListBuilderName, "Previous path found at %s\n", previousHashedVariantListPath.c_str());
+                HashedVariantListSourceData prevHashedVariantList;
+                if (AZ::RPI::JsonUtils::LoadObjectFromFile(previousHashedVariantListPath, prevHashedVariantList, 10 * 1024 * 1024))
```

Comment:
There should be some constant with an explanation for 10 * 1024 * 1024

Reply:
This code has been removed in the latest commit.